src/LeaseManager.cc

```diff
@@ -0,0 +1,307 @@
+ * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL AUTHORS BE LIABLE FOR
+ * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ */
+
+#include <string>
+
+#include "LeaseManager.h"
+
+#include "ExternalStorage.h"
+#include "LeaseCommon.h"
+#include "ShortMacros.h"
+
+namespace RAMCloud {
+
+/// Defines the number of leases the preallocator should try to keep ahead of
+/// the lastIssuedLeaseId.
+const uint64_t PREALLOCATION_LIMIT = 1000;
+
+/// Defines the prefix for objects stored in external storage by this module.
+const std::string STORAGE_PREFIX = "leaseManager";
+
+LeaseManager::LeaseManager(Context* context)
+    : mutex()
+    , context(context)
+    , clock(context)
+    , lastIssuedLeaseId(0)
+    , maxAllocatedLeaseId(0)
+    , leaseMap()
+    , expirationOrder()
+    , preallocator(context, this)
+    , cleaner(context, this)
+{}
+
+/**
+ * Return the lease info for the provided leaseId.  Used by masters to
+ * implicitly check if a lease is still valid.
+ *
+ * \param leaseId
+ *      Id of the lease whose liveness you wish to check.
+ * \return
+ *      If the lease exists, return lease with its current lease term.  If not,
+ *      the returned the lease id will be 0 (an invalid lease id) signaling that
+ *      the requested lease has already expired.
+ */
+WireFormat::ClientLease
+LeaseManager::getLeaseInfo(uint64_t leaseId)
+{
+    Lock lock(mutex);
+    WireFormat::ClientLease clientLease;
+
+    // Populate the lease information if it is found.
+    LeaseMap::iterator leaseEntry = leaseMap.find(leaseId);
+    if (leaseEntry != leaseMap.end()) {
+        clientLease.leaseId = leaseId;
+        clientLease.leaseTerm = leaseEntry->second;
+    } else {
+        clientLease.leaseId = 0;
+        clientLease.leaseTerm = 0;
+    }
+
+    clientLease.timestamp = clock.getTime();
+
+    return clientLease;
+}
+
+/**
+ * Recover lease information from external storage.  This should only be called
+ * once after the construction of the Lease Manager.
+ */
+void
+LeaseManager::recover()
+{
+    Lock lock(mutex);
+
+    // Fetch all lease information from external storage.
+    vector<ExternalStorage::Object> objects;
+    context->externalStorage->getChildren(STORAGE_PREFIX.c_str(), &objects);
+
+    // We can safely make the approximation that every allocated lease was
+    // issued with the exception of the largest allocated leaseId.
+    foreach (ExternalStorage::Object& object, objects) {
+        try {
+            std::string name = object.name;
+            uint64_t leaseId = std::stoull(name);
+
+            // Keep track of the largest allocated leaseId and only auto "renew"
+            // the other leaseIds.
```

Comment:
This seems like unnecessary extra complexity: why not renew even the largest allocated leaseId, even though you don't need to, just to avoid special cases in the code? You're already going to be renewing some leases that have never actually been granted, so  one more won't make much difference.